### PR TITLE
Private heat map

### PR DESCRIPTION
**Pull request by pratiksha**

### CODE REVIEW

File: platform/src/main/java/org/hillview/table/PrivacySchema.java [hunk 1]
```diff
@@ -23,6 +26,17 @@ public PrivacyMetadata get(String colName) {
         return metadata.get(colName);
     }
 
+    public PrivacyMetadata get(String[] colNames) {
```
Comment (mbudiu-vmw) [inline platform/src/main/java/org/hillview/table/PrivacySchema.java]: Can you please add a test with the private metadata that uses multiple columns?
Including serialization/deserialization, etc.

File: web/src/main/java/org/hillview/dataStructures/PrivateHistogram.java [hunk 2]
```diff
@@ -19,25 +21,51 @@
...
         this.confMaxes = new double[this.bucketDescription.getNumOfBuckets()];
+        this.epsilon = epsilon;
+
         this.addDyadicLaplaceNoise();
         if (cdf) {
             this.recomputeCDF();
         }
     }
 
+    // Compute noise to add to this bucket using the dyadic decomposition as the PRG seed.
```
Comment (mbudiu-vmw) [inline web/src/main/java/org/hillview/dataStructures/PrivateHistogram.java]: Please make this a `/**` comment.

File: web/src/main/java/org/hillview/dataStructures/PrivateHistogram.java [hunk 3]
```diff
@@ -19,25 +21,51 @@
...
+    // If cdfBuckets is true, computes the noise based on the dyadic decomposition of the interval [0, bucket right leaf]
```
Comment (mbudiu-vmw) [inline web/src/main/java/org/hillview/dataStructures/PrivateHistogram.java]: There is no 'cdfBuckets'. But you can add documentation for each parameter.

File: web/src/main/java/org/hillview/targets/PrivateTableTarget.java [hunk 4]
```diff
@@ -89,31 +90,33 @@ public void getSchema(RpcRequest request, RpcRequestContext context) {
...
-        }
     }
 
     // compute CDF on the second histogram (at finer CDF granularity)
     private static BiFunction<Pair<Histogram, Histogram>,
             HillviewComputation,
-            Pair<PrivateHistogram, PrivateHistogram>> makePrivateFunction() {
-        return (e, c) -> new Pair<PrivateHistogram, PrivateHistogram>(
-                new PrivateHistogram(e.first, false), new PrivateHistogram(e.second, true));
+            Pair<PrivateHistogram, PrivateHistogram>> makePrivateFunction(double epsilon) {
```
Comment (mbudiu-vmw) [inline web/src/main/java/org/hillview/targets/PrivateTableTarget.java]: I have been inlining this function, it's not really doing much.

File: web/src/main/webapp/dataViews/tableView.ts [hunk 5]
```diff
@@ -636,8 +637,14 @@ export class TableView extends TSViewBase implements IScrollTarget, OnNextK {
                     action: () => this.privateHistSelected(),
                     help: "Plot the data in the selected columns as a private histogram. " +
                         "Applies to one numeric column only.",
-                }, selectedCount === 1);
-            }
+                }, selectedCount === 1 && this.getSelectedColNames().reduce((a, b) => a && this.isNumericColumn(b), true));
```
[the remark this answers is not in the document]
Comment (pratiksha) [reply web/src/main/webapp/dataViews/tableView.ts]: I am going to leave this alone for now until the relevant functionality (string histograms, 2-col histograms if we want to support that) is actually implemented.

**Comment by pratiksha** [comment]
@mbudiu-vmw ptal!

**Pull request merged**